core/script.runtime/src/kotlin/script/templates/annotations.kt

```diff
@@ -22,13 +22,28 @@ import kotlin.reflect.KClass
 import kotlin.script.dependencies.DependenciesResolver.NoDependencies
 import kotlin.script.dependencies.ScriptDependenciesResolver
 
+// discuss
+
+// One of the requests is to allow to share compiler args via script templates: https://youtrack.jetbrains.com/issue/KT-19120
+// Making it a property of ScriptTemplateDefinition is a viable options but it doesn't allow to changes compiler arguments based on script content.
+// For example, it could be possible to allow file level annotation and use it like:
+// @file: ExtraCompilerArgs("-enableWhatever")
+
 const val DEFAULT_SCRIPT_FILE_PATTERN = ".*\\.kts"
 
+// classes annotated with ScriptTemplateDefinition become script templates.
+// examples: https://github.com/JetBrains/kotlin/blob/5faad493b4cf7bf33bf82475e966a99a8e835720/compiler/tests/org/jetbrains/kotlin/scripts/ScriptTemplateTest.kt#L564-L601
 @Target(AnnotationTarget.CLASS)
 @Retention(AnnotationRetention.RUNTIME)
 annotation class ScriptTemplateDefinition(val resolver: KClass<out ScriptDependenciesResolver> = NoDependencies::class,
                                           val scriptFilePattern: String = DEFAULT_SCRIPT_FILE_PATTERN)
 
+// usage example: https://github.com/JetBrains/kotlin/blob/88652154c96402475d42ae0496ab0b423cc0a2b2/libraries/tools/kotlin-script-util/src/main/kotlin/org/jetbrains/kotlin/script/util/resolve.kt#L42-L42
```

Comment:
Seems we're talking about two things here:
1. The laziness of the annotation creation
2. Additional filter in the form of `@AcceptedAnnotations`

I generally disagree with the "pessimization" approach, especially when the optimized one is already in place. Usually, nobody meets real performance problem in one clearly visible bottleneck, but the problem is usually distributed evenly. Skipping the job that nobody needs is usually quite straightforward and implementation here is quite simple. 
On the other hand, if the laziness will become problematic (as @pTalanov mentioned above), the second mechanism of filtering will eliminate the need of constructing unneeded annotations (and in most cases will eliminate the construction altogether, since there are not many annotation-processing resolvers yet written).
Another consideration is that the annotation creation code (https://github.com/JetBrains/kotlin/blob/master/compiler/frontend.script/src/org/jetbrains/kotlin/script/scriptAnnotationsPreprocessing.kt#L35) is not simple, and is relying on some frontend magic, and may need some more logic on to of what is already written, so I don't think that the performance is something we can easily discount here.
And finally, the filtering annotation could be used as an additional type check, ensuring that the resolver receiving the annotations it is able to process.